File: tools/dev/console.py
"""Terminal color support: a single global enable flag plus style helpers.

dev.py and the helper modules print plain text by default; wrapping a message in
one of the style helpers here colors it only when color is enabled, and returns
it unchanged otherwise — so call sites read the same in either mode.

`configure` resolves the mode once at startup (from --plain/--colored or, in auto
mode, from the terminal and the NO_COLOR/FORCE_COLOR conventions); every later
helper reads that one decision. Zero dependencies: just ANSI SGR escapes.
"""

from __future__ import annotations

import os
import sys

_GREEN = "\033[32m"
_YELLOW = "\033[33m"
_RED = "\033[31m"
_DIM = "\033[2m"
_BOLD = "\033[1m"
_RESET = "\033[0m"

_enabled = False


def configure(mode: str = "auto") -> None:
    """Set whether style helpers emit ANSI codes. `mode` is one of:

    - "colored": always on; "plain": always off — the explicit --colored/--plain
      flags, which override everything.
    - "auto": NO_COLOR (any value) forces off and wins over FORCE_COLOR (the
      no-color.org convention); FORCE_COLOR forces on even when piped; otherwise
      on only when both stdout and stderr are TTYs. Requiring both means piping
      either stream (the agent / `| cat` case) yields plain output and keeps ANSI
      out of redirected stdout data.
    """
    global _enabled
    if mode == "colored":
        _enabled = True
    elif mode == "plain":
        _enabled = False
    else:  # auto
        if os.environ.get("NO_COLOR") is not None:
            _enabled = False
        elif os.environ.get("FORCE_COLOR") is not None:
            _enabled = True
        else:
            _enabled = sys.stdout.isatty() and sys.stderr.isatty()


def enabled() -> bool:
    return _enabled


def _wrap(code: str, s: str) -> str:
    return f"{code}{s}{_RESET}" if _enabled else s


def green(s: str) -> str:
    return _wrap(_GREEN, s)


def yellow(s: str) -> str:
    return _wrap(_YELLOW, s)


def red(s: str) -> str:
    return _wrap(_RED, s)


def dim(s: str) -> str:
    return _wrap(_DIM, s)


def bold(s: str) -> str:
    return _wrap(_BOLD, s)

File: tools/instruction-tracer/src/instruction-tracer/report/console.cc
#include "console.hh"

#include <clean-core/platform/win32_sanitized.hh>
#include <clean-core/string/format.hh>
#include <io.h>
#include <stdio.h>

#include <cstdlib>

namespace itrace
{
namespace
{
constexpr cc::string_view reset = "\033[0m";

bool g_enabled = false;

bool is_terminal(FILE* stream)
{
    return _isatty(_fileno(stream)) != 0;
}

/// Old conhost needs to be told it speaks ANSI; Windows Terminal already does. Failure is fine —
/// it only means we are not on a VT-capable console, and the caller's isatty check already gates us.
void enable_windows_vt()
{
    auto* const handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == nullptr || handle == INVALID_HANDLE_VALUE)
        return;

    DWORD mode = 0;
    if (!GetConsoleMode(handle, &mode))
        return;

    SetConsoleMode(handle, mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
}

cc::string wrap(cc::string_view code, cc::string_view s)
{
    if (!g_enabled)
        return cc::string(s);

    return cc::format("{}{}{}", code, s, reset);
}
} // namespace

void configure_console(color_mode mode)
{
    switch (mode)
    {
    case color_mode::always:
        g_enabled = true;
        break;

    case color_mode::never:
        g_enabled = false;
        break;

    case color_mode::automatic:
        // NO_COLOR wins over FORCE_COLOR (the no-color.org convention). Requiring both streams to be
        // terminals means piping either one yields plain output, keeping ANSI out of redirected data.
        if (std::getenv("NO_COLOR") != nullptr)
            g_enabled = false;
        else if (std::getenv("FORCE_COLOR") != nullptr)
            g_enabled = true;
        else
            g_enabled = is_terminal(stdout) && is_terminal(stderr);
        break;
    }

    if (g_enabled)
        enable_windows_vt();
}

bool color_enabled()
{
    return g_enabled;
}

cc::string dim(cc::string_view s)
{
    return wrap("\033[2m", s);
}
cc::string bold(cc::string_view s)
{
    return wrap("\033[1m", s);
}
cc::string red(cc::string_view s)
{
    return wrap("\033[31m", s);
}
cc::string green(cc::string_view s)
{
    return wrap("\033[32m", s);
}
cc::string yellow(cc::string_view s)
{
    return wrap("\033[33m", s);
}
cc::string cyan(cc::string_view s)
{
    return wrap("\033[36m", s);
}
} // namespace itrace
